### DISCUSSION TITLE
Semantic Commit Type of Groups Based on First Dependency in Group

### DISCUSSION THREAD

**Discussion by Kurt-von-Laven** [renovatebot/renovate#18834] [Request Help] [answered]
We are using a [`matchPaths`](https://github.com/ScribeMD/.github/blob/4adb9c4eab74c1ef659e253a097949239cae4589/default.json#L42-L45) package rule to configure a `semanticCommitType` of `fix`, but as you can see from ScribeMD/pre-commit-hooks#132, the `semanticCommitType` is `chore`. Matching `action.yaml` in the same package rule has the expected effect of prefixing the commit message with `fix` as evidenced by ScribeMD/pre-commit-action#164. Is there something wrong with our `matchPaths` pattern given that we are simply trying to match [`.pre-commit-hooks.yaml`](https://github.com/ScribeMD/pre-commit-hooks/blob/28a1a0e1b8896b240d8201ac0aa5a5f76cfef8e9/.pre-commit-hooks.yaml) at the root of a repository?

**Comment by viceice** [comment]
better use `matchFiles`

**Reply by Kurt-von-Laven** [reply]
Good suggestion, but unfortunately it didn't have any effect on the symptom.

**Reply by viceice** [reply]
ok. so we need a minimal reproduction to debug this further.

**Reply by Kurt-von-Laven** [reply]
I have been trying to create one for you in [Kurt-von-Laven/pre-commit-hooks](https://github.com/Kurt-von-Laven/pre-commit-hooks), but the Dependency Dashboard isn't being created. I am not expecting an onboarding PR, because I already have a minimal Renovate config in there. I tried uninstalling and reinstalling Forking Renovate as well as deleting and recreating the repository. Are you able to see anything in the logs?

**Reply by rarkins** [reply]
I checked and there's a backend bug causing the initial forking of the repo to fail. I've attempted to resolve this manually and will also be fixing the root cause soon

**Reply by rarkins** [reply]
btw you can test this in a non-forking app too

**Reply by rarkins** [reply]
Now working: https://github.com/Kurt-von-Laven/pre-commit-hooks/issues/1

But please note I see it has package files detection in 4 different managers so this does not appear to be minimal

**Reply by Kurt-von-Laven** [reply]
Thank you! Yeah, I have been taking a binary search-inspired approach to these minimal reproductions, so the plan was always to iterate based on the presence or absence of the bug. The issue does indeed appear to be with matching a hidden file, because when I match [`README.md` instead of `.pre-commit-hooks.yaml`](https://github.com/Kurt-von-Laven/pre-commit-hooks/commit/1e2c6770b48a8bcc39b58175cf0e18d3c3f71ec6), then the expected commit type of `fix` is used. With [the current config](https://github.com/Kurt-von-Laven/pre-commit-hooks/blob/4cadd91210faa3fa5137787d500512dbe398777d/.github/renovate.json), the incorrect commit type of `chore` is used though.

**Reply by Kurt-von-Laven** [reply]
Would it be possible to convert this discussion to an issue now that it has a minimal reproduction of a confirmed bug?

**Reply by rarkins** [reply]
There's more work to be done to make the config minimal.

For example, 
- why does this need a regex manager to demonstrate a suspected package rule problem?
- is config:base and dependencyDashboardApproval necessary?

Eventually, the README needs to be updated to describe expected vs observed behavior, so that it's clear.

**Reply by Kurt-von-Laven** [reply]
> why does this need a regex manager to demonstrate a suspected package rule problem?

That is a great question. I don't know, but I was unable to reproduce the issue without a regex manager. See Kurt-von-Laven/pre-commit-hooks@1a633a0663e09b5a1aa6b62b51997db66214b847 for a similar config file without a regex manager that does not exhibit the bug. Consequently, I reverted this change.

> is config:base and dependencyDashboardApproval necessary?

No, I only left these in as a convenience to avoid the need to wait ~4 hours between each experiment. I have removed them now though to demonstrate that the bug remains present.

I am not at my desk at the moment, but I will update the README of the minimal reproduction within the hour.

**Reply by rarkins** [reply]
"matchPaths does not match hidden files if a regex manager is present" does not sound right. It sounds like we're missing something else and confusing them.

Anyway the README says:

<img width="830" alt="image" src="https://user-images.githubusercontent.com/6311784/203236104-da1ebcae-97de-438a-83d0-413b1f11cc12.png">

While I see the PR with fix:

<img width="539" alt="image" src="https://user-images.githubusercontent.com/6311784/203236060-2629d020-9c5b-4fc6-b076-d75fb03bef5b.png">

**Reply by Kurt-von-Laven** [reply]
That makes me suspect that the bug is stateful and that some of my experiments have been invalid. As you can see, [Renovate didn't change the PR title until 5:12 a.m. UTC today](https://github.com/Kurt-von-Laven/pre-commit-hooks/pull/3#event-7863885310), 6 hours after I pushed Kurt-von-Laven/pre-commit-hooks@afe302b1e41b777bcae0f9189d4f5c5b149959ba, which was the last config change at the time Renovate ran. Usually it runs when I push to it, and occasionally I have to wait up to a maximum of 3 hours. Is there a way as a user to distinguish between the cases where Forking Renovate did and didn't run on an up-to-date repository? I have reverted the removal of `config:base` to see if that reintroduces the bug, although I am not sure how exactly to wipe whatever state may be confounding these experiments.

**Reply by rarkins** [reply]
If you move your reproduction to the renovate app instead of forking renovate the you can view all logs at app.renovatebot.com.

**Reply by Kurt-von-Laven** [reply]
I switched the reproduction from Forking Renovate to Mend Renovate, but https://app.renovatebot.com/dashboard is blank. I perpetually see a loading indicator on the left in the repos section. I tried uninstalling and reinstalling Mend Renovate to no avail.

**Reply by rarkins** [reply]
Can you check developer tools in your browser to see if there's an error there? It's meant to load like this:

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/6311784/203481166-ec9e63bf-6d35-401b-add2-bae6684b15b1.png">

I'd also recommend you try logging out then in again, or if that's not possible then removing all cookies for the site so that you're logged out.

**Reply by Kurt-von-Laven** [reply]
Thank you for the tips! Indeed there was an error in the developer console, and logging out and back in addressed the issue. I made a version of the reproduction that uses the Mend Renovate app at [Kurt-von-Laven/minimal-reproduction](https://github.com/Kurt-von-Laven/minimal-reproduction), but I haven't found anything in the logs that would explain why the wrong commit type is used.

**Reply by Kurt-von-Laven** [reply]
[Changing the dependency type from `devDependencies` to `dependencies`](https://github.com/Kurt-von-Laven/minimal-reproduction/commit/5c286a9cd9627a4cbf1035d3d01b1b428eea2e13) [appears to have resolved the issue](https://github.com/Kurt-von-Laven/minimal-reproduction/pull/2#event-7881425965). I am not sure why [matching on a non-hidden file instead](https://github.com/Kurt-von-Laven/pre-commit-hooks/commit/1e2c6770b48a8bcc39b58175cf0e18d3c3f71ec6) as well as temporarily [removing](https://github.com/Kurt-von-Laven/pre-commit-hooks/commit/1a633a0663e09b5a1aa6b62b51997db66214b847) and then [re-adding](https://github.com/Kurt-von-Laven/pre-commit-hooks/commit/4c9c6db4d61d3c2be389c30902b20d67ae9ef0a6) the regex manager also worked around the issue. All of this makes me suspect that when the dependency type and `packageRules` contradict one another, either may be used, possibly based on the outcome of a race condition.

**Reply by Kurt-von-Laven** [reply]
The original bug does still reproduce when bumping [a group containing both `devDependencies` and `dependencies`](https://github.com/Kurt-von-Laven/minimal-reproduction/blob/3b0016db3b42e3f4a7d3ab1d2ff43cd3cb99fcb5/.github/renovate.json) though. See Kurt-von-Laven/minimal-reproduction#4, which uses chore instead of fix.

**Reply by rarkins** [reply]
> The original bug does still reproduce when bumping [a group containing both `devDependencies` and `dependencies`](https://github.com/Kurt-von-Laven/minimal-reproduction/blob/3b0016db3b42e3f4a7d3ab1d2ff43cd3cb99fcb5/.github/renovate.json) though. See [Kurt-von-Laven/minimal-reproduction#4](https://github.com/Kurt-von-Laven/minimal-reproduction/pull/4), which uses chore instead of fix.

This is to be expected. When groups have a mix of semantic commit types, the result is non-deterministic.

**Reply by rarkins** [reply]
I think you'll find this is caused by the config:base preset which assigns fix/chore based on dependency type. If so then there's no bug

**Reply by Kurt-von-Laven** [reply]
> This is to be expected. When groups have a mix of semantic commit types, the result is non-deterministic.

Non-deterministic behavior doesn't seem very desirable. Ideally, Renovate would use the "highest" commit type, but since that can't possibly be determined in the general case, even something completely arbitrary like using the first commit type lexicographically would be preferable. It takes significantly more time and effort to create minimal reproductions when one cannot assume the determinism of the system under test. For whatever reason, our Commitizen group, which also mixes development and production dependencies has thus far always used fix.

> I think you'll find this is caused by the config:base preset which assigns fix/chore based on dependency type. If so then there's no bug

Yes, that makes sense; I did recently check that the issue couldn't be reproduced without `extends: ["config:base"]`. Do you recommend specifying the commit type in the same package rule that defines the group?

**Reply by viceice** [reply]
It's partially deterministic: renovate uses the options from the first update in group.

**Reply by rarkins** [reply]
It's deterministic in the sense that it takes the first upgrade's semantic commit type, and there's rules for how the upgrades are sorted, but are you suggesting just sorting them alphabetically?

I recommend you either add an ignore for the nested preset `:semanticPrefixFixDepsChoreOthers` within `config:base`, or you mimic its structure by writing it using a package rule

**Reply by Kurt-von-Laven** [reply]
We can probably work with that. How does Renovate determine the order of the updates within a group?

**Reply by rarkins** [reply]
Sorts by update type: pin, digest, patch, minor, major

Within those they're sorted by the order they're found, which should be manager name and then package file name

**Reply by Kurt-von-Laven** [reply]
Thank you; that is helpful to know. Using the first upgrade's commit type seems preferable to using the first commit type sorted alphabetically. It would be ideal if certain well known commit types, such as those in Renovate-provided presets had a precedence to them (e.g., fix > chore), and the "highest" commit type of any dependency applied to groups. We would ideally like to use fix for commits that bump both the mega-linter-runner npm package and also the MegaLinter Docker image, yet chore for commits that only bump the MegaLinter Docker image. Unfortunately, the package filename for the Docker image is `.pre-commit-config.yaml`, while the package filenames for the npm package are `.pre-commit-hooks.yaml` and `README.md`. I don't think adding `:semanticPrefixFixDepsChoreOthers` to `ignorePresets` will allow us to achieve this behavior. If I understand correctly, the only option would be to stop grouping MegaLinter bumps?

**Reply by rarkins** [reply]
Generally speaking Renovate can't understand every type of commit prefix and decide which is more important. However `fix` vs `chore` is a very common one we could maybe hardcode a rule for.

**Reply by Kurt-von-Laven** [reply]
That would be excellent. Those are the only ones I'm aware of there currently being Renovate presets for, so that seems like it would be a convenient line to draw in the sand. This seems obvious in retrospect, but we can simply have a special package rule in the one repository where MegaLinter bumps should be fixes rather than chores. The thing we can't do currently is control the commit type of grouped dependencies as a function of which subset of the dependencies in the group are bumped in a given PR.

**Comment by Kurt-von-Laven** [comment, marked as answer]
To summarize, the semantic commit type for the first dependency in the group is used when the package rule for the group itself doesn't explicitly specify a `semanticCommitType`. Dependencies are ordered deterministically as described [here](https://github.com/renovatebot/renovate/discussions/18834#discussioncomment-4222571).  We were able to address our issue by specifying a semantic commit type for the group in one repository and disabling grouping to other repositories where it didn't apply to avoid overriding the commit type there. Big thanks for all of the debugging support; that was a long one.
